### PR TITLE
L2 support miner change

### PR DESCRIPTION
**Pull request by syntrust**
For the miner to read Randao from L1 blocks, add a polling client to sync the latest L1 block number known by the L2 by calling `L1Block` contract.

### CODE REVIEW

File: ethstorage/eth/polling_client.go [hunk 1]
```diff
@@ -164,20 +144,27 @@ func (w *PollingClient) pollHeads() {
...
+			num, err := w.GetLatestNumber()
 			if err != nil {
-				w.lgr.Info("error getting latest header", "err", err)
+				w.lgr.Info("error getting latest l1 block number", "err", err)
 				reqPollAfter()
 				continue
 			}
-			if w.currHead != nil && w.currHead.Hash() == head.Hash() {
-				w.lgr.Trace("no change in head, skipping notifications")
+			if w.currL1Number != nil && w.currL1Number.Cmp(num) == 0 {
```
Comment (qzhodl) [inline ethstorage/eth/polling_client.go:153]: how do we handle the reorg case here

Comment (syntrust) [reply ethstorage/eth/polling_client.go]: back to compare header hash

File: ethstorage/eth/polling_client.go [hunk 2]
```diff
@@ -191,20 +185,42 @@ func (w *PollingClient) pollHeads() {
 	}
 }
 
-func (w *PollingClient) getLatestHeader() (*types.Header, error) {
-	ctx, cancel := context.WithTimeout(w.ctx, 5*time.Second)
+func (w *PollingClient) GetLatestHeader() (*types.Header, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
```
Comment (qzhodl) [inline ethstorage/eth/polling_client.go:189]: Should we use the passed w.ctx instead of a new one?

File: ethstorage/eth/randao_client.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+var l1BlockContract = common.HexToAddress("0x4200000000000000000000000000000000000015")
+
+// RandaoClient is a client that fetches the latest randao from the L1 block contract.
+// It uses the PollingClient to fetch the latest block number and then fetches the header
+// corresponding to that block number.
+type RandaoClient struct {
+	*PollingClient
+	// rpc endpoint where randao is fetched from
+	rc      *ethclient.Client
+	rCtx    context.Context
```
Comment (qzhodl) [inline ethstorage/eth/randao_client.go:29]: It seems that we don't use this context

**Review by qzhodl — APPROVED** [review]

**Review by qzhodl — APPROVED** [review]

**Review by ping-ke — APPROVED** [review]

**Pull request merged**